timemachine/fe/dummy.py

```diff
@@ -91,19 +102,18 @@ def identify_root_anchors(mol, core, dummy_atom):
         List of root anchors that the dummy atom is connected to.
 
     """
-
     assert len(set(core)) == len(core)
 
     core = set(core)
-
     assert dummy_atom not in core
 
-    # first convert to a dense graph
-    N = mol.GetNumAtoms()
+    # first convert to a dense graph, and assume that bond_idxs start from zero
+    flat_idxs = np.array(bond_idxs).flatten()
+    N = np.amax(flat_idxs) + 1
+    assert set(flat_idxs) == set(range(N))
     dense_graph = np.zeros((N, N), dtype=np.int32)
 
-    for bond in mol.GetBonds():
-        i, j = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
+    for i, j in bond_idxs:
```

Comment:
tbh I don't understand the APIs on the page `https://networkx.org/documentation/stable/reference/algorithms/traversal.html` well enough right now to try and convert this, some of them many of the functionalities are for direct graphs. I'm trying to avoid too much logic change right now (and just make this a clean-up PR). Let me read up on this a bit more to understand how to use it?

Reply:
oh, sorry, I wasn't necessarily suggesting to switch to networkx for the implementation immediately; just that refactoring this slightly might make for a clearer separation between the bespoke parts of the algorithm vs. the more standard graph traversal implementation (which could be swapped out later)

Reply:
After this some more thought (and attempting to refactor this into network graphs and using `nx.dfs_tree`), this is not doable as a standard `dfs` algorithm. The issue here is that if we induce a graph that's comprised of dummy atoms (and no core atoms), we have no way to actually "appending a core atom" to the anchor list, because it simply doesn't exist! No matter what dfs we do on the induced graph, it will only ever contain dummy atoms. There is no way to "terminate on first visit of a core atom".

Reply:
Resolved after offline discussion.